File: backend/pipeline/downloader.py
"""
PDF downloader for government document URLs.
Downloads with browser-like headers, validates MIME type, saves to temp dir.
"""

import hashlib
import logging
import tempfile
from pathlib import Path

import httpx

logger = logging.getLogger(__name__)

_HEADERS = {
    "User-Agent": (
        "Mozilla/5.0 (Windows NT 10.0; Win64; x64) "
        "AppleWebKit/537.36 (KHTML, like Gecko) "
        "Chrome/124.0.0.0 Safari/537.36"
    ),
    "Accept": "application/pdf,*/*",
}

_TIMEOUT = httpx.Timeout(connect=10, read=60, write=30, pool=5)
_MAX_RETRIES = 3


class DownloadError(Exception):
    pass


def _temp_path(url: str) -> Path:
    name = hashlib.sha256(url.encode()).hexdigest()[:16]
    return Path(tempfile.gettempdir()) / f"{name}.pdf"


async def download_pdf(url: str) -> Path:
    """Download a PDF from *url* and return the local temp path.

    Retries up to _MAX_RETRIES times with exponential backoff.
    Raises DownloadError on failure or non-PDF response.
    """
    import asyncio

    dest = _temp_path(url)
    if dest.exists():
        logger.debug("Cache hit for %s → %s", url, dest)
        return dest

    last_exc: Exception | None = None
    for attempt in range(_MAX_RETRIES):
        try:
            async with httpx.AsyncClient(
                headers=_HEADERS, timeout=_TIMEOUT, follow_redirects=True
            ) as client:
                resp = await client.get(url)
                resp.raise_for_status()

            content_type = resp.headers.get("content-type", "")
            if "application/pdf" not in content_type and not url.lower().endswith(".pdf"):
                raise DownloadError(
                    f"Expected application/pdf, got '{content_type}' for {url}"
                )

            dest.write_bytes(resp.content)
            logger.info("Downloaded %s → %s (%d bytes)", url, dest, len(resp.content))
            return dest

        except DownloadError:
            raise
        except (httpx.HTTPError, OSError) as exc:
            last_exc = exc
            if attempt < _MAX_RETRIES - 1:
                wait = 2 ** attempt
                logger.warning("Download attempt %d failed (%s), retrying in %ds", attempt + 1, exc, wait)
                await asyncio.sleep(wait)

    raise DownloadError(f"Failed to download {url} after {_MAX_RETRIES} attempts: {last_exc}") from last_exc

File: frontend/lib/pipeline/downloader.ts
/**
 * PDF downloader — port of backend/pipeline/downloader.py
 */

import { createHash } from 'crypto'
import { tmpdir } from 'os'
import { join } from 'path'
import { readFile, writeFile, access } from 'fs/promises'
import { BROWSER_HEADERS } from './types'

const MAX_RETRIES = 3
const TIMEOUT_MS = 60_000

function tempPath(url: string): string {
  const name = createHash('sha256').update(url).digest('hex').slice(0, 16)
  return join(tmpdir(), `${name}.pdf`)
}

export async function downloadPdf(url: string): Promise<Buffer> {
  const dest = tempPath(url)

  // Return cached file if it exists
  try {
    await access(dest)
    return readFile(dest)
  } catch {
    // not cached
  }

  let lastError: unknown
  for (let attempt = 0; attempt < MAX_RETRIES; attempt++) {
    try {
      const resp = await fetch(url, {
        headers: { ...BROWSER_HEADERS, Accept: 'application/pdf,*/*' },
        signal: AbortSignal.timeout(TIMEOUT_MS),
        redirect: 'follow',
      })
      if (!resp.ok) throw new Error(`HTTP ${resp.status}`)

      const contentType = resp.headers.get('content-type') ?? ''
      if (!contentType.includes('application/pdf') && !url.toLowerCase().endsWith('.pdf')) {
        throw new Error(`Expected application/pdf, got '${contentType}'`)
      }

      const buf = Buffer.from(await resp.arrayBuffer())
      await writeFile(dest, buf)
      return buf
    } catch (err) {
      lastError = err
      if (attempt < MAX_RETRIES - 1) {
        await new Promise(r => setTimeout(r, 2 ** attempt * 1000))
      }
    }
  }
  throw new Error(`Failed to download ${url} after ${MAX_RETRIES} attempts: ${lastError}`)
}
